### ISSUE TITLE
Secure Boot with Encrypted Flash Not Working (IDFGH-1251)

### ISSUE THREAD

**Issue by rhr407** [espressif/esp-idf#3548]
## Environment

- Development Kit: [ESP32-DEVKIT]
- Kit version : [v1]
- Module or chip used: [ESP32-WROOM-32|ESP32-WROOM-32D|ESP32-WROOM-32U]
- Chip: ESP32D0WDQ6
- IDF version (run ``git describe --tags`` to find it): 
    ``` v4.0-dev-621-g81ca1c013``` 
- Build System:         [Make|CMake]
- Compiler version (run ``xtensa-esp32-elf-gcc --version`` to find it):
``` 
xtensa-esp32-elf-gcc (crosstool-NG crosstool-ng-1.22.0-80-g6c4433a) 5.2.0
Copyright (C) 2015 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
``` 
- Operating System:     [Linux]
- Power Supply:         [USB]

## Problem Description
The problem is that on terminal the wroom32 dev kit is continuously printing the following message: 
``` 
rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
flash read err, 1000
ets_main.c 371 
ets Jun  8 2016 00:22:57
``` 
Below are the screenshots of terminal that I hope would be relevant information regarding this problem.
//Detailed problem description goes here.
The make menuconfig out put is given below: 
![Security-features](https://user-images.githubusercontent.com/26026993/58517129-48f1c500-81a2-11e9-827a-837251ee849d.png)

![Security-Unsafe-options](https://user-images.githubusercontent.com/26026993/58517128-48f1c500-81a2-11e9-8b87-c3897ea4a507.png)

![Partition-table](https://user-images.githubusercontent.com/26026993/58517131-48f1c500-81a2-11e9-8654-a6319767e829.png)

### Expected Behavior
I do not know the expected behavior as I have never used secure boot with encrypted flash before.
### Actual Behavior
![message](https://user-images.githubusercontent.com/26026993/58517130-48f1c500-81a2-11e9-8ab3-1f6cf3a19668.png)

### Steps to repropduce
I have been struggling very hard to get the secure boot and flash encryption to work for last three days.
The issue is that secure boot works fine without flash encryption but when I enable the flash encryption it gives me error. Ironically, the flash encryption also work fine without secure boot. So separately they both work well. Let me mention the method of enabling the Secure Boot first and then I will write the steps for flash boot.

###SECURE BOOT:

1. Give the command on terminal:
``` 
$ make menuconfig
``` 
2. Follow the settings in the image given below or in above section.
![image_1](https://user-images.githubusercontent.com/26026993/58518740-675abf00-81a8-11e9-88c4-024f721d1ccf.png)

3. Exit and save settings
4. Give the following command in the terminal:
``` 
$ openssl ecparam -name prime256v1 -genkey -noout -out secure_boot_signing_key.pem 
``` 
(This will generate a .pem file inside the project folder)
5. Now, give the following command in terminal:
``` 
$ make && make erase-flash && make bootloader
``` 
6. As a result of step 5 above, two sets of flashing steps will be printed - the first set of steps includes an espefuse.py burn_key command which is used to write the bootloader key to efuse. (Flashing this key is a one-time-only process.) The second set of steps can be used to reflash the bootloader with a pre-calculated digest (generated during the build process). Run both these set of commands.
7. Now run:
``` 
$ make flash monitor
``` 
The program should boot now in secure mode.

8. Please do check the efuse condition using this command:
``` 
$ espefuse.py .summary
``` 

You should see the similar output on terminal after this command

![espefuse py-summary](https://user-images.githubusercontent.com/26026993/58518780-7e011600-81a8-11e9-99a8-562f22576d12.png)

As you can see that BLK1 and BLK2 keys (for Flash encryption and Secure Boot) have already been burnt in my case. You should see 00 00 00 .... before burning these keys

As you can see the ABS_DONE_0 is 1 so it means that the secure boot is enabled.

One thing that made me annoyed so much was this efuse burning option which is irreversible. Espressif documentation does not warn in red flag that the user must be very careful in burning these fuses. I have ended up in bricking all the 3 or my esp32 dev v1 kits. 2 of them I ordered on Amazon and got on Friday. But all the 3 of them are bricked now.

###FLASH ENCRYPTION

1. Give the command on terminal:
``` 
$ make menuconfig
``` 
2. Follow the settings in the image_3

![image_3](https://user-images.githubusercontent.com/26026993/58518795-8d805f00-81a8-11e9-861e-d282b8901807.png)

3. Save Settings and exit
4. Give the following command in terminal to generate a flash encryption key:
```  
$ espsecure.py digest_private_key --keyfile secure_boot_signing_key.pem --keylen 256 my_flash_encryption_key.bin
``` 
5. Burn the generated key in BLK_1 block for flash encryption using the following command: 
``` 
$ espefuse.py --port /dev/ttyUSB0 burn_key flash_encryption my_flash_encryption_key.bin
``` 
6. Now, give the following command in terminal:
``` 
$ make && make erase_flash flash monitor
``` 
On startup, the flash will be encrypted and then bootloader will boot from flash.

###OTA and Secure Boot

1. In order to use the OTA using secure boot, follow the steps fro SECURE BOOT for the OTA app to sign the .bin file inside the build folder using the .pem key file generated in step 4 of SECURE BOOT heading.

2. Use the same .pem file for signing an OTA app. For this purpose copy and paste the .pem file inside the OTA app folder and build using make command. The signed .bin file for app will be generated using this .pem key.

When the factory app will retrieve the OTA app from remote server, it will verify the signed binary digest. If newly fetched OTA binary was signed using the same .pem key the bootloader will boot the app, otherwise it will not.

I have ended up in bricking my 3 devkits and hence stopped further working. I have search through esp-idf github issues quite a lot but could not find the solution for resolve the bricking issue. I want to now submit a new issue regarding the simultaneous use of secure boot and flash encryption on the github as I end up having a loop message like given in image_4.

![message](https://user-images.githubusercontent.com/26026993/58518868-d0dacd80-81a8-11e9-92c7-08d092ca41b7.png)

Please do check the efuse condition using this command ---> $ espefuse.py .summary

Kindly help me resolve this issue and advise me if I have done any step wrong as mentioned above.
### Code to reproduce this issue

```cpp
// the code should be wrapped in the ```cpp tag so that it will be displayed better.
#include "esp_log.h"

void app_main()
{
    
}

```
// If your code is longer than 30 lines, [GIST](https://gist.github.com) is preferred.

## Debug Logs

```
Debug log goes here, should contain the backtrace, as well as the reset source if it is a crash.
Please copy the plain text here for us to search the error log. Or attach the complete logs but leave the main part here if the log is *too* long.
```

## Other items if possible

- [X] sdkconfig file (attach the sdkconfig file from your project folder)
- [x] elf file in the ``build`` folder (**note this may contain all the code details and symbols of your project.**)
- [ ] coredump (This provides stacks of tasks.)

**Comment by Alvin1Zhang** [comment]
Thanks for reporting the issue. Would you please help provide more details as suggested in the [issue template](https://github.com/espressif/esp-idf/blob/master/docs/issue_template.md)? Thanks.

**Comment by renzbagaporo** [comment]
Also, the problem description seem to be just the issue template copied verbatim and do not actually contain details about the issue. We encourage filling this template out so that we can help you as best as we could.

**Comment by negativekelvin** [comment]
You can't erase the bootloader after you have previously enabled secure boot.  You also can't enable flash encryption after you have previously enabled secure boot, you have to enable them both together. 

I do agree that there should be more explicit warnings like "Production Only! Not for development use!" and a development mode that doesn't read protect the efuse keys.

**Comment by rhr407** [comment]
So it means if I have to use the Secure boot with flash encryption I have to enable both simultaneously first? 
and then follow the key burning step
then make bootloader step and
then make flash monitor
All the above steps after I have generated the .pem key file

Please correct me if I am wrong.

**Comment by ghemal** [comment]
Sorry to hear about your problem and thanks for the detailed information

The continuous boot loop indicates the ROM code can not understand the flash contents when it tries to load the second stage boot loader from flash. This can happen due to wrong configuration (such as flash encryption is enabled but flash contents are not encrypted or vice versa). Also once enabled it is not possible to disable secure boot in device (even if you disable it through menuconfig)

In your case it seems you have enabled secure boot and flash encryption individually one at a time and then tried to disable secure boot and then flash the device (perhaps bootloader got programmed multiple times with different configuration). If you want to use both the features it is recommended to enable them together to avoid any misconfiguration issue. However, good thing I see here is you have selected secure boot mode as “re-flashable" and flash encryption key is generated "externally". So you have access to both the keys (secure boot & flash encryption) to create new bootable image.

I would suggest you to try following steps and let us know if it helps

1. In menuconfig ensure secure boot mode as “reflashable” as well as flash encryption is enabled
2. make bootloader, this will generate a bootloader binary with secure digest
3. Encrypt the bootloader binary using flash encryption key with encrypt_flash_data command. For example please refer [.](https://docs.espressif.com/projects/esp-idf/en/latest/security/flash-encryption.html#reflashing-with-pregenerated-key) The bootloader is written at address 0x1000
4. Next you can program the newly encrypted bootloader binary using the same command as was printed by make bootloader (but instead use encrypted bootloader bin)
5. Similarly encrypt the partition table binary and application binary with same flash encryption key using appropriate addresses in your case (usually the partition table resides at 0x8000 and default app image at 0x10000). If you have OTA partitions do the same for them
6. Program all the encrypted partitions in above step using write_flash command and corresponding address for each partition. Sample command

> esptool.py --port /dev/ttyUSB0 --baud 115200 write_flash 0x10000 build/my-app.bin

7. Reset the device

**Comment by rhr407** [comment]
@ghemal : Thank you so much for your detailed reply. @gh@gh

I have followed the above steps in the following manner:

## BEFORE ENCRYPTING THE PARTITIONS:

1. 
```   
make menuconfig
```  

2. Set the following:
a. Secure boot and flash encryption steps in menuconfig:
![Screenshot-20190605122946-554x225](https://user-images.githubusercontent.com/26026993/58953194-a6cb7180-878d-11e9-9585-6c8ac27e6c79.png)
b. Set two OTA partitions:
![Screenshot-20190605123202-557x142](https://user-images.githubusercontent.com/26026993/58953311-f6aa3880-878d-11e9-978c-7800da210f0a.png)

Above are the outputs in my case.

3. Do the Key Generation steps for secureboot and flash encryption putting the secure_boot_signing_key.pem and my_flash_encryption_key.bin in the project folder.
```
openssl ecparam -name prime256v1 -genkey -noout -out my_secure_boot_signing_key.pem
```
```
espsecure.py digest_private_key --keyfile secure_boot_signing_key.pem --keylen 256 my_flash_encryption_key.bin
```

4. 
``` 
make clean 
``` 

5. To make it sure that we do not have any other data in flash:
``` 
make erase_flash
``` 

5. This following command will generate the `app.bin` and `partitions_two_ota.bin` files 
``` 
make -j4
``` 

6. Generate bootloader.bin file
``` 
make bootloader
``` 

will generate the `bootlaoder.bin` file inside build/bootloader/bootloader.bin folder

7. Check addresses of partitions you have in generated partition table:

``` 
make partition_table
``` 
![Screenshot-20190605121718-564x139](https://user-images.githubusercontent.com/26026993/58952549-eb560d80-878b-11e9-979b-5c352ed09b55.png)
The above is the output in my case.

The above command will show the addresses for each partition in generated partition table before using in the following section:

## ENCRYPTING ALL THE PARTITIONS  (generated above) AND BURNING IN THE FLASH:

1. First encrypt the `bootloader.bin` file using the my_flash_encryption_key.bin file at addres 0x1000:

```  
espsecure.py encrypt_flash_data --keyfile my_flash_encryption_key.bin --address 0x1000 -o build/bootloader/bootloader-encrypted.bin build/bootloader/bootloader.bin 
```  

2. Burn the encrypted `bootloader-encrypted.bin` in the flash using the following command:
``` 
python esp-idf/components/esptool_py/esptool/esptool.py --chip esp32 --port /dev/ttyUSB0 --baud 921600 --before default_reset --after hard_reset write_flash -z --flash_mode dio --flash_freq 40m --flash_size detect 0x1000 /build/bootloader/bootloader-encrypted.bin

``` 


3. Now encrypt the `partitions_two_ota.bin` using the my_flash_encryption_key.bin file at addres 0x9000:
``` 
espsecure.py encrypt_flash_data --keyfile my_flash_encryption_key.bin --address 0x9000 -o build/partitions_two_ota-encrypted.bin build/partitions_two_ota.bin
``` 
4. Burn the encrypted `partitions_two_ota-encrypted.bin` in the flash using the following command:
``` 
esptool.py --port /dev/ttyUSB0 --baud 921000 write_flash 0x9000 build/partitions_two_ota-encrypted.bin
``` 

5. Next encrypt the `app.bin` file at address 0x10000:
```
espsecure.py encrypt_flash_data --keyfile my_flash_encryption_key.bin --address 0x10000 -o build/app-encrypted.bin build/app.bin
```
6. Burn the `app-encrypted.bin` into the flash:
``` 
esptool.py --port /dev/ttyUSB0 --baud 921000 write_flash 0x10000 build/app-encrypted.bin
```

After doing ``make monitor `` , the screen is continuously printing as follows:
![Screenshot-20190605120735-499x275](https://user-images.githubusercontent.com/26026993/58952087-c01eee80-878a-11e9-9387-5b913d945dc5.png)


NOTE: I have not yet encrypted the OTA partitions as I just have to first test the app

**Comment by rhr407** [comment]
Got new ESP32 DOIT devkit v1 
Did all the above steps. I have noticed that we have to burn the 'espefuse.py burn_efuse FLASH_CRYPT_CNT' and 'ABS_DONE_0' fuses manually. Just enabling the option for secure boot and flash encryption do nothing. I have tested this.
![Screenshot-20190605135527-731x699](https://user-images.githubusercontent.com/26026993/58958098-0cbdf600-879a-11e9-9bac-c07c378c0d17.png)


![Screenshot-20190605135527-731x699](https://user-images.githubusercontent.com/26026993/58957929-a0db8d80-8799-11e9-9b54-53d0aa7816be.png)

**Comment by ghemal** [comment]
Hi rhr407,

Thanks for the details

The steps I recommended were for the device which already has flash encryption and secure boot enabled and when you want to recover the device with improper image. They wouldn't work on a _fresh_ device where flash encryption and secure boot is not yet enabled (for which the original documentation of flash encryption and secure boot needs to be followed)

I take it from your log that in your case you put the encrypted bootloader on a device which hasn't enabled flash encryption (as you can see from your efuse summary) so obviously it can't boot

You have manually programmed the FLASH_CRYPT_CNT and ABS_DONE_0. In addition for flash encryption to work you also need to program FLASH_CRYPT_CONFIG to 0xF as per [this](https://docs.espressif.com/projects/esp-idf/en/latest/security/flash-encryption.html#setting-flash-crypt-config). Ideally these efuses gets programmed when the bootloader runs for the first time (after enabling these features) but since in your case bootloader couldn't execute the efuses never got programmed

Two additional things to check are

1. I see that your default factory app is at offset 0x20000 as per partition table but you are programming the app.bin at 0x10000 which has phy init data so this needs to be corrected

2. Please check the size of your bootloader.bin. There is one known problem where ESP32 fails to boot when secure boot and flash encryption are enabled together as mentioned [here](https://docs.espressif.com/projects/esp-idf/en/latest/security/secure-boot.html#bootloader-size)

**Comment by rhr407** [comment]
Thank you so much for your message. I really appreciate your help and assistance. Also thanks for highlighting my mistake of not burning flash at 0x20000 (that is the address of the factory app in my partition table)

I repeated all the steps you mentioned in your previous message. I did the following steps:
1. Showing the status of efuses:
```  
espefuse.py summary
```  
![Screenshot-20190610214200-733x705](https://user-images.githubusercontent.com/26026993/59225437-9e799900-8bc8-11e9-81d0-0221b7b7f31d.png)

2. Compiling the app and generating bootloader:
```  
make clean && make erase_flash && make -j4 && make bootloader
```  
![Screenshot-20190610211147-785x771](https://user-images.githubusercontent.com/26026993/59223664-62443980-8bc4-11e9-955e-6207efbeb6ae.png)

Here two binaries have been generated. bootloader.bin and bootloader-reflash-digest.bin files.

3. Check for the generated partition tables addresses:
```  
make partition_table
```  
![Screenshot-20190610212248-784x220](https://user-images.githubusercontent.com/26026993/59224282-eba83b80-8bc5-11e9-8c1a-900c0a499cee.png)


4. Signing and writing the .bin files in to the flash:
   a. Signing bootloader.bin as follows with my_flash_encryption_key.bin file [NOTE**:I need to know if I have to do this step for bootloader-reflash-digest.bin as well??]:
      ```  
      espsecure.py encrypt_flash_data --keyfile my_flash_encryption_key.bin --address 0x1000 -o build/bootloader/bootloader-encrypted.bin build/bootloader/bootloader.bin
      ```  
      ![Screenshot-20190610211411-786x59](https://user-images.githubusercontent.com/26026993/59223781-b7804b00-8bc4-11e9-80f4-63c051b1ee1a.png)

   b. Writing the encrypted bootloader to flash at 0x1000:
   ```  
   esptool.py --port /dev/ttyUSB0 --baud 921000 write_flash 0x1000 build/bootloader/bootloader-encrypted.bin
   ```  
   ![Screenshot-20190610211707-672x294](https://user-images.githubusercontent.com/26026993/59223983-29589480-8bc5-11e9-8bb7-596c8a9c1209.png)

   c. Now encrypting partitions_singleapp.bin:
   ```  
   espsecure.py encrypt_flash_data --keyfile my_flash_encryption_key.bin --address 0x12000 -o build/partitions_singleapp-encrypted.bin build/partitions_singleapp.bin
   ```
   ![Screenshot-20190610213132-784x73](https://user-images.githubusercontent.com/26026993/59224902-5f971380-8bc7-11e9-9a0b-f8290993f1a4.png)

   d. Writing the app-encrypted.bin to flash at 0x12000
   ```    
   esptool.py --port /dev/ttyUSB0 --baud 921000 write_flash 0x12000 build/partitions_singleapp-encrypted.bin
   ``` 

   ![Screenshot-20190610213448-708x294](https://user-images.githubusercontent.com/26026993/59225012-9cfba100-8bc7-11e9-8516-0479274cd4e6.png)

   e. Now encrypting app.bin: [It will take a while as the app size is bigger]
   ```  
   espsecure.py encrypt_flash_data --keyfile my_flash_encryption_key.bin --address 0x20000 -o build/app-encrypted.bin build/app.bin
   ```  

   ![Screenshot-20190610212649-699x55](https://user-images.githubusercontent.com/26026993/59224507-7852f980-8bc6-11e9-8183-32f795a33af1.png)

   f. Writing the app-encrypted.bin to flash at 0x20000
   ```  
   esptool.py --port /dev/ttyUSB0 --baud 921000 write_flash 0x20000 build/app-encrypted.bin
   ``` 

   ![Screenshot-20190610212804-713x290](https://user-images.githubusercontent.com/26026993/59224574-a6d0d480-8bc6-11e9-84ed-71b6a2d36284.png)

5. Reset the esp32 and watch monitor:
```  
make monitor
```  

![Screenshot-20190610213935-501x155](https://user-images.githubusercontent.com/26026993/59225308-43e03d00-8bc8-11e9-92da-8dbcbd0a2a51.png)


NOW I am facing this issue. The above image shows the terminal output continuously printing the following text:

rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0018,len:4
load:0x3fff001c,len:10444
load:0x40078000,len:18888
load:0x40080400,len:6580
secure boot check fail
ets_main.c 371 
ets Jun  8 2016 00:22:57

**Comment by negativekelvin** [comment]
Yes you have to write the encrypted bootloader digest to 0x0 or the combined encrypted bootloader-digest.bin to 0x0.

**Comment by rhr407** [comment]
Awesome!! 
It's working now Alhamdulillah! 

![Screenshot-20190610232331-742x769](https://user-images.githubusercontent.com/26026993/59230885-ddaee680-8bd6-11e9-88a8-81e0b90029a8.png)

Thank you so much for your kind assistance and support.

**Comment by chhotu** [comment]
Hi,
I have enabled Secure boot(Reflashable mode) + flash encryption(Development mode).
I have built 'Simple_ota_example' and able to run properly.

Now, I want to change Secure boot mode to 'one-time-flash' from 'Reflashable'
Can anyone help me, which steps I need to follow for the same?

Note: I want to switch this configuration in the same board.


1, Reflashable Mode configuration : 
![Development_Mode](https://user-images.githubusercontent.com/50351479/62039384-3a14a680-b215-11e9-955c-45a999276fe7.JPG)

2. one-time flash mode configuration : 
![Release_Mode](https://user-images.githubusercontent.com/50351479/62039433-544e8480-b215-11e9-9e43-9c02e23f2af9.JPG)


Apart from this what kind of steps do I need to perform?

Replay would be appreciated !!!!
